row_cache.cc

```diff
@@ -78,7 +96,35 @@ cache_tracker::cache_tracker(mutation_application_stats& app_stats, register_met
                 return memory::reclaiming_result::reclaimed_something;
             }
             current_tracker = this;
-            return _lru.evict();
+
+            // Cache replacement algorithm:
+            //
+            // if sstable index caches occupy more than index_cache_fraction of cache memory:
+            //     evict the least recently used index entry
+            // else:
+            //     evict the least recently used entry (data or index)
+            //
+            // This algorithm has the following good properties:
+            // 1. When index and data entries contend for cache space, it prevents
+            //    index cache from taking more than index_cache_fraction of memory.
+            //    This makes sure that the index cache doesn't catastrophically
+            //    deprive the data cache of memory in small-partition workloads.
+            // 2. Since it doesn't enforce a lower limit on the index space, but only
+            //    the upper limit, the parameter shouldn't require careful balancing.
+            //    In workloads where it makes sense to cache the index (usually: where
+            //    the index is small enough to fit in RAM), setting the fraction to any big number (e.g. 1.0)
+            //    should work well enough. In workloads where it doesn't make sense to cache the index,
+            //    setting the fraction to any small number (e.g. 0.0) should work well enough.
+            //    Setting it to a medium number (something like 0.2) should work well enough
+            //    for both extremes, although it might be suboptimal for non-extremes.
+            // 3. The parameter is trivially live-updateable.
+            //
+            // Perhaps this logic should be encapsulated somewhere else, maybe in `class lru` itself.
+            size_t total_cache_space = _region.occupancy().total_space();
+            size_t index_cache_space = sstables::space_used_by_index_cache();
+            bool should_evict_index = index_cache_space > total_cache_space * _index_cache_fraction.get();
+
+            return _lru.evict(should_evict_index);
```

Comment:
Might want to loop here to amortize the cost of the decision.

Reply:
The decision should be very cheap, amortizing it isn't worth the added complexity.

Reply:
It may be cheap but the gain from an eviction can be tiny, so we have a low-cost decision against a low-gain outcome.

Reply:
The thing is, I expect eviction of an item to be orders of magnitude more expensive than computing the conditional. The minimum cost of a single eviction is DRAM latency (it's an uncached linked list, after all), and usually it also entails unlinking stuff from trees, destroying cells, freeing memory, etc. Compared to that, computing the conditional should be rather insignificant — it's a few cacheable loads and a bit of arithmetic.

Reply:
Yes probably true re. the uncacheability and its dominance.